lib/charms/prometheus_k8s/v0/prometheus.py

```diff
@@ -573,48 +579,91 @@ def _labeled_static_job_config(self, job, job_name_prefix, hosts, scrape_metadat
+                    )
                 )
-                config["static_configs"].append(unitless_config)
 
-            for host_name, host_address in hosts.items():
-                static_config = self._labeled_unit_config(
-                    host_name, host_address, ports, labels, scrape_metadata
-                )
-                config["static_configs"].append(static_config)
-                if "juju_unit" not in relabel_config["source_labels"]:
-                    relabel_config["source_labels"].append("juju_unit")
+            if ports := [
+                target.split(":")[1] for target in targets if target.startswith("*:")
+            ]:
+                if self._is_aggregated_static_config(static_config, scrape_metadata):
+                    # Do not add targets for proxy applications that publish scrape
+                    # jobs with Juju labels that do not match the scrape metadata
+                    logger.error(
+                        "Received a static_config that uses the '*:<port>' target "
+                        "schema, and that has Juju topology labels that do not match "
+                        "the scrape_metadata"
+                    )
+                    continue
 
-        config["relabel_configs"] = [relabel_config]
+                for host_name, host_address in hosts.items():
+                    static_configs.append(
+                        self._labeled_unit_config(
+                            host_name, host_address, ports, labels, scrape_metadata
+                        )
+                    )
 
-        return config
+        return {
+            "job_name": job_name,
+            "metrics_path": job["metrics_path"],
+            "relabel_configs": [
+                # This relabelling rule will match metrics that do not have the full Juju topology,
+                # and ensure that metrics without a juju_unit are still unique per instance
+                {
+                    "source_labels": [
+                        "juju_model",
+                        "juju_model_uuid",
+                        "juju_application",
+                        "__address__",
+                    ],
+                    "separator": "_",
+                    "target_label": "instance",
+                    "regex": "(.*)",
+                },
+                # This relabelling rule will match all metrics with the full juju topology,
+                {
+                    "source_labels": [
+                        "juju_model",
+                        "juju_model_uuid",
+                        "juju_application",
+                        "juju_unit",
+                    ],
+                    "separator": "_",
+                    "target_label": "instance",
+                    "regex": "(.*)",
+                },
+            ],
+            "static_configs": static_configs,
+        }
+
+    def _is_aggregated_static_config(self, static_config, scrape_metadata):
```

Comment:
On second thoughts here is an alternative that may be worth considering : `scrape_metadata` is passed only by `PrometheusConsumer` since it is used for creating job names and labels. However if a different object such as `PrometheusAggregator` may be free to take an entirely different approach, leaving `scrape_metadata` only for use by individual charms. For example `PrometheusAggregator` can submit the entire job spec and be wholly responsible for setting job names, labels etc. Then the absence of `scrape_metadata` or if required an alternative key say `aggregated_metadata` can be indicative of what `PrometheusProvider` needs to deal with. With this approach the changes to `PrometheusProvider` are minimal, with little cross talk between the two cases (such as checking for intersecting labels), and no changes to `PrometheusConsumer`. 

Finally there remains the case of aggregating alert rules. I have not had the opportunity to think this through but my instinct is to by keeping `PrometheusAggregator` as a separate object from `PrometheusConsumer` this use case will also be easy to solve.

Reply:
> But what happens if a errant charm accidentally does set a Juju topology label or intentionally wants to ?

Accidental: Bug
Intentional: Feature

Reply:
> could have an incomplete set of topology labels.

This is worth testing for!